django/test/selenium.py

```diff
@@ -191,14 +193,28 @@ def dark(self):
             finally:
                 self.selenium.execute_script("localStorage.removeItem('theme');")
 
-    def set_emulated_media(self, features, media=""):
-        if self.browser != "chrome":
-            self.skipTest("Emulated media controls are only supported on Chrome.")
+    def set_emulated_media(self, *, media=None, features=None):
+        if self.browser not in {"chrome", "edge"}:
+            self.skipTest("Emulated media controls are only supported on Chrome/Edge.")
+
         # Chrome Dev Tools Protocol Emulation.setEmulatedMedia
         # https://chromedevtools.github.io/devtools-protocol/1-3/Emulation/#method-setEmulatedMedia
-        self.selenium.execute_cdp_cmd(
-            "Emulation.setEmulatedMedia", {"media": media, "features": features}
-        )
+
+        # Using the CDP API to ensure this works with remote drivers for Selenium Grid.
+        # https://www.selenium.dev/documentation/webdriver/bidirectional/chrome_devtools/cdp_api/
+
+        import trio  # Trio is a dependency of Selenium.
+
+        async def execute_cdp_async():
+            async with self.selenium.bidi_connection() as connection:
+                emulation = connection.devtools.emulation
+                command = emulation.set_emulated_media(
+                    media=media,
+                    features=[emulation.MediaFeature(**x) for x in features],
+                )
+                return await connection.session.execute(command)
+
+        trio.run(execute_cdp_async)
```

Comment:
I was attempting to make `set_emulated_media()` work consistently when using either standalone Selenium or Selenium Grid (which is used in `django-docker-box`; see my fork for a working version).

Alas, while this works in both, it doesn't seem to be having the desired effect of enabling forced colors properly... I'll need to look into this some more. Maybe it's a race condition or something... 😞

Reply:
I think I've managed to fix this by taking a different approach 🙂